hypervisor/libvirt/libvirt.go

```diff
@@ -0,0 +1,699 @@
+	return string(data), nil
+}
+
+func scsiId2Name(id int) string {
+	return "sd" + utils.DiskId2Name(id)
+}
+
+func scsiId2Addr(id int) (int, int, error) {
+	if id > 65535 {
+		return -1, -1, fmt.Errorf("id %d too long, exceed 256*256")
+	}
+
+	return id / 256, id % 256, nil
+}
+
+func (lc *LibvirtContext) AddDisk(ctx *hypervisor.VmContext, name, sourceType, filename, format string, id int) {
+	diskXml, err := diskXml(filename, format, id)
+	if err != nil {
+		glog.Error("generate attach-disk-xml failed, ", err.Error())
+		ctx.Hub <- &hypervisor.DeviceFailed{
+			Session: nil,
+		}
+		return
+	}
+	glog.V(3).Infof("diskxml: %s", diskXml)
+
+	err = lc.domain.AttachDeviceFlags(diskXml, libvirtgo.VIR_DOMAIN_DEVICE_MODIFY_LIVE)
+	if err != nil {
+		glog.Error("attach disk device failed, ", err.Error())
+		ctx.Hub <- &hypervisor.DeviceFailed{
+			Session: nil,
+		}
+		return
+	}
+	target, unit, err := scsiId2Addr(id)
+	ctx.Hub <- &hypervisor.BlockdevInsertedEvent{
+		Name:       name,
+		SourceType: sourceType,
+		DeviceName: scsiId2Name(id),
+		ScsiId:     id,
+		ScsiAddr:   fmt.Sprintf("%d:%d", target, unit),
+	}
+}
+
+func (lc *LibvirtContext) RemoveDisk(ctx *hypervisor.VmContext, filename, format string, id int, callback hypervisor.VmEvent) {
+	diskXml, err := diskXml(filename, format, id)
+	if err != nil {
+		glog.Error("generate detach-disk-xml failed, ", err.Error())
+		ctx.Hub <- &hypervisor.DeviceFailed{
+			Session: callback,
+		}
+		return
+	}
+	err = lc.domain.DetachDeviceFlags(diskXml, libvirtgo.VIR_DOMAIN_DEVICE_MODIFY_LIVE)
+	if err != nil {
+		glog.Error("detach disk device failed, ", err.Error())
+		ctx.Hub <- &hypervisor.DeviceFailed{
+			Session: callback,
+		}
+		return
+	}
+	ctx.Hub <- callback
+}
+
+type nicmac struct {
+	Address string `xml:"address,attr"`
+}
+
+type nicsrc struct {
+	Bridge string `xml:"bridge,attr"`
+}
+
+type nictgt struct {
+	Device string `xml:"dev,attr,omitempty"`
+}
+
+type nicmodel fsdriver
+
+type nic struct {
+	XMLName xml.Name `xml:"interface"`
+	Type    string   `xml:"type,attr"`
+	Mac     nicmac   `xml:"mac"`
+	Source  nicsrc   `xml:"source"`
+	Target  *nictgt  `xml:"target,omitempty"`
+	Model   nicmodel `xml:"model"`
+	Address *address `xml:"address"`
+}
+
+func nicXml(bridge, device, mac string, addr int) (string, error) {
+	slot := fmt.Sprintf("0x%x", addr)
+	n := nic{
+		Type: "bridge",
+		Mac: nicmac{
+			Address: mac,
+		},
+		Source: nicsrc{
+			Bridge: bridge,
+		},
+		Target: &nictgt{
+			Device: device,
+		},
+		Model: nicmodel{
+			Type: "virtio",
+		},
+		Address: &address{
+			Type:     "pci",
+			Domain:   "0x0000",
+			Bus:      "0x00",
+			Slot:     slot,
+			Function: "0x0",
+		},
+	}
+
+	data, err := xml.Marshal(n)
+	if err != nil {
+		return "", err
+	}
+	return string(data), nil
+}
+
+func (lc *LibvirtContext) AddNic(ctx *hypervisor.VmContext, host *hypervisor.HostNicInfo, guest *hypervisor.GuestNicInfo) {
+	nicXml, err := nicXml(host.Bridge, host.Device, host.Mac, guest.Busaddr)
```

Comment:
Sorry for taking you into the trap, here should be `host.Device` indeed. Network nic will be set up correctly after merging #95 and reverting this change.